spec.md

```diff
@@ -1225,28 +1391,94 @@ module MyModule :
-component will have its value automatically sign-extended or zero-extended to
-the larger bit width. The behavior of connect statements between two
-circuit components with aggregate types is defined by the connection algorithm
-in [@sec:the-connection-algorithm].
+To ensure connections are meaningful when taking directionality into account,
+every expression in FIRRTL has a flow.
+
+The flow of an expression can be one of *source*, *sink*, or *duplex*.
+
+A source expression supplies a signal, and can be used to drive a circuit component.
+A sink expression can be driven by another expression.
+A duplex expression is simply an expression that is both a source and sink.
+
+The rules for the flow of an expression are as follows:
+
+If the expression is a reference, we look at the kind of the circuit component:
+
+* Nodes are sources.
+* Wires and registers are duplex.
+* For ports, `input` ports are sources and `output` ports are sinks.
+* Submodule instances are sources.
+* Memories are sources.
+
+Here are a few comments to help with intuition:
+Nodes may only appear "on the left side" of a connect.
+Wires and registers may appear "on either side of a connect statement".
+Ports are always considered from the perspective of "inside the module".
+Moreover, input ports may only appear "on the left side" of a connect,
+while output ports may only appear "on the right side" of a connect.
+Finally, while submodules instances and memories are strictly sources,
+they interact with the sub-field rule below, allowing connections to their input ports.
+
+If the expression is a sub-field expression, the flow depends on whether the field is flipped.
+If the field is not flipped, the flow is the same as the expression itself.
+If the field is flipped, the flow is the reversed flow:
+
+* The reverse of a source is a sink.
+* The reverse of a sink is a source.
+* The reverse of a duplex is a duplex.
+
+All remaining expressions are sources:
+
+* constant literals
+* primitive operations
+* multiplexers
+* enum expressions
+* all probe expressions (`probe` `rwprobe` and `read`)
+
+
+## Connection Rules
 
-### The Connection Algorithm
+Knowing both the type and flow of both sides of a `connect` statement,
+you can determine whether the connection is legal.
 
-Connect statements between ground types cannot be expanded further.
+The types of the left-hand and right-hand side expressions must be
+equivalent (see [@sec:type-equivalence] for details).
+
+The flow of the right-hand side expression must be source or duplex,
+or else the right-hand side expression has a passive type.
+
+TODO: Why is a passive type acceptable?
+TODO: Is this because passive types "can be flipped for free" intuitively?
+
+The flow of the left-hand side expression must be sink or duplex.
+
+Neither side of the connect may have a property type.
+
+TODO: I think this is fraught. What if a property ends up in a bundle type?
```

Comment:
@mikeurbach I'm looking for some help with this. I think the intention is that property types can never enter bundles. That seems to be suggestive in the grammar.